Create ls.rb

ngmysh opened the pull request:
lsコマンドを作成

ngmysh committed:
Create ls.rb

lsコマンドを作成

u1tnk commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,65 @@
+
+COLUMN = 3
+
+def pad_divide(array, number, fill_with)
```
こちらtransposeを前提に配列操作しているのでtransposeと不可分だと思うので、ここにtransposeも一緒に入れて良いと思います。
名前を変える必要があると思うので適切な名前を考えてください 🙏

u1tnk commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,65 @@
+def print_ls_style(entries)
+  return if entries.empty?
+
+  # entriesを各列の配列に分割する。
```
コメントは変数名、メソッド名を工夫してどうしても伝わりそうに無いときに付けるようにしましょう。ちょっと詳し過ぎると思いました。
pad_divideを名前を変えて欲しいとコメントしているので合わせて、何をやっているか分かる名前にして、このコメントは簡潔にしてメソッドの方につけましょう。

u1tnk commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,65 @@
+  # 後でtransposeでエラーにならないように、entriesが3で割り切れないときは最後の配列は空文字列パディングする。
+  divided_entries = pad_divide(entries.sort, COLUMN, '')
+
+  # 列の表示をそろえるために、各列の中で最大の文字列の長さを記録しておく。
```
ここはコメント削除して良いです `max_length_each_column`という名前が適切なので推測可能なので不要と思います 👍

u1tnk commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,65 @@
+  entries = Dir.glob('*')
+  print_ls_style(entries)
+else
+  # 引数でファイルやディレクトリが指定された場合、処理される順番は次のようになっていた。（ls (GNU coreutils) 8.32）
```
このコメントも詳し過ぎますので、この流れがメソッド名なので分かるようにできるととても良いですね 💪 
単に消して良い部分もありますが、全部消すとちょっと分かりづらいので、まずは最低限のコメントにして、可能であればメソッド化等でコメント消せるように頑張ってみましょう 💪

ngmysh replied:
この部分は、メソッドに分ける方法がなかなか思いつかず、コメントを最小限にすることで対応しました💦

u1tnk reviewed (changes requested):
業務のプログラミングは原則としてソースコードでできるだけ表現する。コメントは補助という関係があるので、そこを見直して欲しいです。

実装はメソッド化も使えているし、全体的に整理されて見やすいので上記以外は大丈夫そうです 👍

ngmysh committed:
Clean some code to readable

無駄なコメントを削除し、メソッドを追加することでソースコードを読みやすくした

u1tnk commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,68 @@
+  formatted_entries = format_entries(entries, COLUMN)
+  max_length_each_column = formatted_entries.transpose.map { |column| column.max_by(&:length).size }
+
+  print_aligned_entries(formatted_entries, max_length_each_column)
```
`print_aligned_entries` → `print_entries` にして上の二行も入れちゃいましょう。
そうすれば 配列と表示する列数を入れれば自動的に表示してくれる便利メソッドになりますね。

すみません、先に言っておけば良かったですが、この状態になってみて一緒にした方がシンプルになると思いまして 🙏

ngmysh committed:
Change print_aligned_entries to print_entries and Fix bug

`print_aligned_entries`メソッドを`print_entries`に変更して役割をスッキリさせた
引数に１つのファイル、１つのディレクトリが指定されたときに、ディレクトリ名が表示されないバグを修正

ngmysh commented:
ご指摘いただいた点と、１つバグを一つ修正しました！
（引数に１つのファイル、１つのディレクトリが指定されたときに、ディレクトリ名が表示されないバグを修正）

u1tnk commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,68 @@
+  end
+end
+
+def show_entries(entries)
```
ここまで来たらこのメソッド不要ですね。 print/showどちらでも良いですが統一しましょう。

u1tnk reviewed (changes requested):
すみません、改善が進んだのでまた一点コメントしました 🙏 
あとreaseするほどでも無いですが、バグ修正とその他の修正は別の場所なのでコミット分けた方が良かったですね。

ngmysh commented:
> バグ修正とその他の修正は別の場所なのでコミット分けた方が良かったですね。

今後気を付けます💦

統一すると、rubocopに`show_entries`メソッドの複雑度が高いと怒られてしまったので、それぞれの列で一番長い文字列の長さを求める部分をメソッド化しました🙌

ngmysh committed:
Change `show_entries` and `print_entries` to just show_entries

`show_entries`と`print_entries`を`show_entries`に統一、各列の最大の文字列の長さを求める処理をメソッド化

u1tnk reviewed (approved):
だいぶシンプルになって良くなったと思います。お疲れ様でした 👍

ngmysh commented:
ありがとうございました～！
mainブランチにマージします
